### PR TITLE
Extract shared children logic when handling components to ComponentSet

### PR DESCRIPTION
**Pull request by luanpotter**
# Description

This unifies how we handle "children" for both *BaseGame* and *BaseComponent* by creating a wrapper over OrderedSet.

This wrapper does all the handling of addLater and removeLater, making sure that BaseComponent does not have any concurrency modifications issues.

It also unifies the APIs, keeping most component list related methods inside the `components` (or `children` namespaces). I do expose a few most important methods, like `add`, specially since add does some extra work as well. But more specific things, like `remove`, `length`, `clear`, I left inside the namespace for clarity. I consider removing even `add` but thought that was too idiomatic `game.add`. But open to suggestions on that area.

I also didn't touch `propagateToChildren` here, nor rebalancing. But I believe those two can be simplified now that both Base* use the same structure. That is just implementation details though, which I might look into next. Also wanna draw a state flowchart to better explain the difference between mount and load (that I had forgotten).

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors and are passing (See [Contributor Guide]).
- [x] My PR does not decrease the code coverage, or I have __a very special case__ and explained on the PR description why this PR decreases the coverage.
- [x] I updated/added relevant documentation (doc comments with `///`) and updated/added examples in `doc/examples`.
- [x] I have formatted my code with `./scripts/format.sh` and the Flame analyzer (`./scripts/analyze.sh`) does not report any problems.
- [x] I read and followed the [Flame Style Guide].
- [x] I have added a description of the change under `[next]` in `CHANGELOG.md`.
- [x] I removed the `Draft` status, by clicking on the `Ready for review` button in this PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (please indicate a breaking change in `CHANGELOG.md`).
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md

### CODE REVIEW

File: examples/lib/stories/components/priority.dart [hunk 1]
```diff
@@ -52,6 +52,6 @@ class Priority extends BaseGame with HasTapableComponents {
       Square(Vector2(170, 150)),
       Square(Vector2(110, 150)),
     ];
-    addAll(squares);
+    components.addAll(squares);
```
Comment (luanpotter) [inline examples/lib/stories/components/priority.dart:55]: addAll is one that was considering if could be added proxied

File: examples/lib/stories/components/composability.dart [hunk 2]
```diff
@@ -27,7 +27,7 @@ class ParentSquare extends Square {
       Square(Vector2(70, 200), Vector2(50, 50), angle: 5),
     ];
 
-    children.forEach(addChild);
+    children.forEach((c) => addChild(c, gameRef: gameRef));
```
Comment (luanpotter) [inline examples/lib/stories/components/composability.dart:30]: this was actually a bug. if any child had a HasGameRef it would not be populated. my new assert catches that.

File: packages/flame/example/lib/main.dart [hunk 3]
```diff
@@ -70,7 +70,7 @@ class MyGame extends BaseGame with DoubleTapDetector, TapDetector {
 
     final handled = components.any((c) {
       if (c is PositionComponent && c.toRect().overlaps(touchArea)) {
-        remove(c);
+        components.remove(c);
```
Comment (luanpotter) [inline packages/flame/example/lib/main.dart:73]: again, another method that I decided not to proxy. open to suggestions

File: packages/flame/lib/src/components/base_component.dart [hunk 4]
```diff
@@ -236,4 +201,12 @@ abstract class BaseComponent extends Component {
   Vector2 eventPosition(PositionInfo info) {
     return isHud ? info.eventPosition.widget : info.eventPosition.game;
   }
+
+  ComponentSet createComponentSet() {
+    final components = ComponentSet.createDefault(prepare);
+    if (this is HasGameRef) {
+      components.register<HasGameRef>();
```
Comment (luanpotter) [inline packages/flame/lib/src/components/base_component.dart:208]: this allows HasGameRef to do `query<HasGameRef>()`, which should improve performance.
also I make all ordered sets QOS by default. so no more complications into whether we can use QOS's features or not

File: packages/flame/lib/src/components/mixins/has_collidables.dart [hunk 5]
```diff
@@ -1,12 +1,9 @@
...
-
 import '../../../game.dart';
 import '../../components/mixins/collidable.dart';
 import '../../geometry/collision_detection.dart';
 
 mixin HasCollidables on BaseGame {
   void handleCollidables() {
-    final qos = components as QueryableOrderedSet;
-    collisionDetection(qos.query<Collidable>());
+    collisionDetection(components.query<Collidable>());
```
Comment (luanpotter) [inline packages/flame/lib/src/components/mixins/has_collidables.dart:7]: everything is QOS now

File: packages/flame/lib/src/components/mixins/has_game_ref.dart [hunk 6]
```diff
@@ -17,9 +17,10 @@ mixin HasGameRef<T extends Game> {
   set gameRef(T gameRef) {
     _gameRef = gameRef;
     if (this is BaseComponent) {
+      // TODO(luan) this is wrong, should be done using propagateToChildren
```
Comment (luanpotter) [inline packages/flame/lib/src/components/mixins/has_game_ref.dart:20]: on a HasGameRef > no game ref > HasGameRef chain this will not work.

I will revisit this when trying to generify propageToChildren later.

File: packages/flame/lib/src/game/base_game.dart [hunk 7]
```diff
@@ -97,14 +83,14 @@ class BaseGame extends Game with FPSCounter {
   ///
   /// You can return a specific sub-class of OrderedSet, like
   /// [QueryableOrderedSet] for example, that we use for Collidables.
-  OrderedSet<Component> createOrderedSet() {
-    final comparator = Comparing.on<Component>((c) => c.priority);
+  ComponentSet createComponentSet() {
+    final components = ComponentSet.createDefault(
+      (c, {BaseGame? gameRef}) => prepare(c),
```
Comment (luanpotter) [inline packages/flame/lib/src/game/base_game.dart:88]: gameRef is always `this` on this case, no need to wire that up

File: packages/flame/lib/src/game/base_game.dart [hunk 8]
```diff
@@ -156,38 +147,8 @@ class BaseGame extends Game with FPSCounter {
   /// This methods is an async operation since it await the `onLoad` method of the component. Nevertheless, this method only need to be waited to finish if by some reason, your logic needs to be sure that the component has finished loading, otherwise, this method can be called without waiting for it to finish as the BaseGame already handle the loading of the component.
   ///
   /// *Note:* Do not add components on the game constructor. This method can only be called after the game already has its layout set, this can be verified by the [hasLayout] property, to add components upon a game initialization, the [onLoad] method can be used instead.
-  Future<void> add(Component c) async {
-    assert(
-      hasLayout,
-      '"add" called before the game is ready. Did you try to access it on the Game constructor? Use the "onLoad" method instead.',
-    );
```
Comment (luanpotter) [inline packages/flame/lib/src/game/base_game.dart:163]: moved to prepare

File: packages/flame/lib/src/game/base_game.dart [hunk 9]
```diff
@@ -156,38 +147,8 @@ class BaseGame extends Game with FPSCounter {
...
-    prepare(c);
-    final loadFuture = c.onLoad();
-
-    if (loadFuture != null) {
-      await loadFuture;
-    }
```
Comment (luanpotter) [inline packages/flame/lib/src/game/base_game.dart:169]: this is done by ComponentSet now

File: packages/flame/lib/src/game/base_game.dart [hunk 10]
```diff
@@ -156,38 +147,8 @@ class BaseGame extends Game with FPSCounter {
...
-  }
-
-  /// Marks a list of components to be removed from the components list on the next game tick
-  void removeAll(Iterable<Component> components) {
-    _removeLater.addAll(components);
-  }
-
-  /// Marks all existing components to be removed from the components list on the next game tick
-  void clear() {
-    _removeLater.addAll(components);
```
Comment (luanpotter) [inline packages/flame/lib/src/game/base_game.dart:190]: proxy cleanup

File: packages/flame_forge2d/lib/forge2d_game.dart [hunk 11]
```diff
@@ -54,15 +53,6 @@ class Forge2DGame extends BaseGame {
     }
   }
 
-  @override
-  void remove(Component component) {
```
Comment (luanpotter) [inline packages/flame_forge2d/lib/forge2d_game.dart:58]: game has no "remove" no more :) 
also this would not work for removeAll before.

File: packages/flame/lib/src/components/mixins/draggable.dart [hunk 12]
```diff
@@ -84,7 +84,7 @@ mixin HasDraggableComponents on BaseGame {
   }
 
   void _onGenericEventReceived(bool Function(Draggable) handler) {
-    for (final c in components.toList().reversed) {
+    for (final c in components.reversed()) {
```
Comment (spydon) [inline packages/flame/lib/src/components/mixins/draggable.dart:87]: I think I might have written this initially, but we should probably just iterate over the list in reverse instead of creating a new reversed list to iterate over.

File: packages/flame/lib/src/game/base_game.dart [hunk 13]
```diff
@@ -156,38 +147,8 @@ class BaseGame extends Game with FPSCounter {
...
+  Future<void> add(Component c) {
+    return components.addChild(c);
```
Comment (spydon) [inline packages/flame/lib/src/game/base_game.dart:151]: I think that if we have `add` here, we should also have the other methods proxied.

**Review by spydon — COMMENTED** [review]
Nicely streamlined with QOS! I do think we should keep the proxies though (or alternatively remove `game.add` too.

File: examples/lib/stories/components/composability.dart (hunk 2 above)
Comment (luanpotter) [reply examples/lib/stories/components/composability.dart:30]: no, but they could

File: packages/flame/lib/src/components/mixins/draggable.dart (hunk 12 above)
Comment (luanpotter) [reply packages/flame/lib/src/components/mixins/draggable.dart:87]: that does exactly the same thing as before

File: examples/lib/stories/components/composability.dart (hunk 2 above)
Comment (spydon) [reply examples/lib/stories/components/composability.dart:30]: But this is an example that the writer of the example is in total control over writing that game? I don't see the need to give the children a `gameRef` when they don't need it?

File: packages/flame/lib/src/game/base_game.dart (hunk 13 above)
Comment (luanpotter) [reply packages/flame/lib/src/game/base_game.dart:151]: I thought about it, but the problem is there are so many collection methods, so where do we draw the line?

File: packages/flame/lib/src/components/mixins/draggable.dart (hunk 12 above)
Comment (spydon) [reply packages/flame/lib/src/components/mixins/draggable.dart:87]: That's what I'm saying, that it was inefficient before too and that we should consider not creating a new list every time an event is fired. So either maintain another list in the reverse order inside of QOS, or iterate it from the end.

File: packages/flame/lib/src/game/base_game.dart (hunk 13 above)
Comment (spydon) [reply packages/flame/lib/src/game/base_game.dart:151]: I would say that the ones we had was a pretty good place to draw the line, but I do see your point.

Comment (spydon) [reply packages/flame/lib/src/game/base_game.dart:151]: Now when I have thought about it a bit more I think I actually prefer the proxies removed, like you have done (possibly adding `addAll` too like you said). Since removal can just as well be done on a component level for single components, and the rest might be clearer if they are on the collection. I think `add` is good to keep though since it is so heavily used.

Comment (luanpotter) [reply packages/flame/lib/src/game/base_game.dart:151]: add, addAll, remove, clear? I thought it could be confusing, specially game.clear vs game.components.clear. but no strong opinion, I can definitely add them again as I said. also the baseComponent didn't expose all those, so we should unify that as well. but let's here what everyone has to say

File: packages/flame/lib/src/components/mixins/draggable.dart (hunk 12 above)
Comment (luanpotter) [reply packages/flame/lib/src/components/mixins/draggable.dart:87]: ah gotcha, let me see how to iterate properly. I think that was the reason I "extracted" to a method but then I forgot to look into it

File: examples/lib/stories/components/composability.dart (hunk 2 above)
Comment (luanpotter) [reply examples/lib/stories/components/composability.dart:30]: that is true, but I thought it would lead to confusing behavior to allow this, might cause not so clever users to get into weird states and think it's a bug

Comment (spydon) [reply examples/lib/stories/components/composability.dart:30]: If that is the case, maybe we should consider having the `gameRef` on all `BaseComponent` as default instead?

Comment (spydon) [reply examples/lib/stories/components/composability.dart:30]: Ah right, that wouldn't solve the case when children are added to the component before the component is added to the game though...

File: packages/flame/lib/src/components/mixins/draggable.dart (hunk 12 above)
Comment (luanpotter) [reply packages/flame/lib/src/components/mixins/draggable.dart:87]: I don't think it's possible, sadly. unless we keep a secondary copy of the entire set in sync with negative priorities. you can't reverse an interator just by the nature of the interface that it exposes. you have to first materialize the whole list

Comment (spydon) [reply packages/flame/lib/src/components/mixins/draggable.dart:87]: Okay, let's handle this later if it becomes a bottle neck.

**Review by spydon — APPROVED** [review]
I think we should add `addAll` and after that I think it looks good!

File: examples/lib/stories/components/composability.dart (hunk 2 above)
Comment (luanpotter) [reply examples/lib/stories/components/composability.dart:30]: I think we can improve that on followups, it's not directly related to this PR

File: packages/flame/lib/src/game/base_game.dart (hunk 13 above)
Comment (luanpotter) [reply packages/flame/lib/src/game/base_game.dart:151]: I will add addAll back, and only that. we can reconsider the api later of course

File: examples/lib/stories/components/priority.dart (hunk 1 above)
Comment (spydon) [reply examples/lib/stories/components/priority.dart:55]: Now when `addAll` is back that one should be used

File: packages/flame/example/lib/main.dart (hunk 3 above)
Comment (spydon) [reply packages/flame/example/lib/main.dart:73]: Maybe we should start doing `c.remove()` in our examples, not sure which one is clearer.

File: packages/flame_forge2d/example/lib/blob_sample.dart [hunk 14]
```diff
@@ -94,7 +94,7 @@ class BlobSample extends Forge2DGame with TapDetector {
     final worldCenter = screenToWorld(size * camera.zoom / 2);
     final blobCenter = worldCenter + Vector2(0, 10);
     final blobRadius = Vector2.all(6.0);
-    addAll(createBoundaries(this));
+    components.addAll(createBoundaries(this));
```
Comment (erickzanardo) [inline packages/flame_forge2d/example/lib/blob_sample.dart:97]: Can we keep addAll here?

File: packages/flame_forge2d/example/lib/sprite_body_sample.dart [hunk 15]
```diff
@@ -46,7 +46,7 @@ class SpriteBodySample extends Forge2DGame with TapDetector {
   @override
   Future<void> onLoad() async {
     _pizzaImage = await images.load('pizza.png');
-    addAll(createBoundaries(this));
+    components.addAll(createBoundaries(this));
```
Comment (erickzanardo) [inline packages/flame_forge2d/example/lib/sprite_body_sample.dart:49]: Can we keep addAll here?

File: packages/flame_forge2d/lib/body_component.dart [hunk 16]
```diff
@@ -133,4 +133,10 @@ abstract class BodyComponent<T extends Forge2DGame> extends BaseComponent
   bool containsPoint(Vector2 point) {
     return body.fixtures.any((fixture) => fixture.testPoint(point));
   }
+
+  @override
+  void onRemove() {
```
Comment (erickzanardo) [inline packages/flame_forge2d/lib/body_component.dart:138]: This seems to be a different fix right? It deserves its own PR, or its own entry on the changelog

File: examples/lib/stories/components/priority.dart (hunk 1 above)
Comment (luanpotter) [reply examples/lib/stories/components/priority.dart:55]: ah I added only to BaseGame. let me add to BaseComp

File: examples/lib/stories/components/composability.dart (hunk 2 above)
Comment (luanpotter) [reply examples/lib/stories/components/composability.dart:30]: we only need to pass when the parent doesn't have it

File: packages/flame/lib/src/components/base_component.dart [hunk 17]
```diff
@@ -236,4 +201,12 @@ abstract class BaseComponent extends Component {
   Vector2 eventPosition(PositionInfo info) {
     return isHud ? info.eventPosition.widget : info.eventPosition.game;
   }
+
+  ComponentSet createComponentSet() {
```
[the remark this answers is not in the document]
Comment (luanpotter) [reply packages/flame/lib/src/components/base_component.dart:205]: yes, you might want to extend it

File: packages/flame_forge2d/lib/body_component.dart (hunk 16 above)
Comment (luanpotter) [reply packages/flame_forge2d/lib/body_component.dart:138]: it's not a fix, just changing the way it's done to work well with this PR

Comment (erickzanardo) [reply packages/flame_forge2d/lib/body_component.dart:138]: Still, at least a changelog entry is needed right, this is on another package.

**Review by erickzanardo — APPROVED** [review]
LGTM once the changelog entry for forged2d is added

File: packages/flame_forge2d/lib/body_component.dart (hunk 16 above)
Comment (luanpotter) [reply packages/flame_forge2d/lib/body_component.dart:138]: I don't see why, there is no functional difference right? just impl detail

Comment (luanpotter) [reply packages/flame_forge2d/lib/body_component.dart:138]: but I can add one
